config/config.go

```diff
@@ -225,6 +207,234 @@ func validateReferencesExist(c configpb.Configuration) error {
+	if tg.TestNameConfig != nil {
+		nameFormat := tg.GetTestNameConfig().GetNameFormat()
+		nameElements := tg.GetTestNameConfig().GetNameElements()
+
+		if len(nameElements) == 0 {
+			mErr = multierror.Append(mErr, &ConfigError{tg.Name, "TestGroup", "TestNameConfig.NameElements must be specified."})
+		}
+
+		if nameFormat == "" {
+			mErr = multierror.Append(mErr, &ConfigError{tg.Name, "TestGroup", "TestNameConfig.NameFormat must be specified."})
+		} else {
+			if got, want := len(nameElements), strings.Count(nameFormat, "%"); got != want {
+				mErr = multierror.Append(
+					mErr,
+					&ConfigError{tg.Name, "TestGroup", fmt.Sprintf("TestNameConfig has %d elements, format %s wants %d", got, nameFormat, want)},
+				)
+			}
+			elements := make([]interface{}, 0)
+			for range nameElements {
+				elements = append(elements, "")
+			}
+			s := fmt.Sprintf(nameFormat, elements...)
+			if strings.Contains(s, "%!") {
+				return &ConfigError{tg.Name, "TestGroup", fmt.Sprintf("Number of format strings and name_elements must match; got %s (%d).", s, len(elements))}
+			}
+		}
+	}
+
+	return mErr.ErrorOrNil()
+}
+
+func validateDashboardTab(dt *configpb.DashboardTab) error {
+	mErr := &multierror.Error{}
+
+	// Check that required fields are a non-zero-value.
+	if dt.TestGroupName == "" {
+		mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "test_group_name can't be empty"})
+	}
+
+	// A Dashboard Tab can't be named the same as the default 'Summary' tab.
+	if dt.Name == "Summary" {
+		mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "tab can't be named 'Summary'"})
+	}
+
+	// TabularNamesRegex should be valid and have capture groups defined.
+	if dt.TabularNamesRegex != "" {
+		regex, err := regexp.Compile(dt.TabularNamesRegex)
+		if err != nil {
+			mErr = multierror.Append(
+				mErr,
+				&ConfigError{dt.Name, "DashboardTab", fmt.Sprintf("Invalid regex %s: %v", dt.TabularNamesRegex, err)})
+		} else {
+			if regex.NumSubexp() != len(regex.SubexpNames()) {
+				mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "all tabular_name_regex capture groups must be named"})
+			}
+			if len(regex.SubexpNames()) < 1 {
+				mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", "tabular_name_regex requires at least one capture group"})
+			}
+		}
+	}
+
+	// Email address for alerts should be valid.
+	if dt.GetAlertOptions().GetAlertMailToAddresses() != "" {
+		err := validateEmails(dt.GetAlertOptions().GetAlertMailToAddresses())
+		if err != nil {
+			mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", err.Error()})
+		}
+	}
+
+	return mErr.ErrorOrNil()
+}
+
+func validateEntityConfigs(c configpb.Configuration) error {
+	mErr := &multierror.Error{}
+
+	// At the moment, don't need to further validate Dashboards or DashboardGroups.
+	for _, tg := range c.TestGroups {
+		err := validateTestGroup(tg)
+		if err != nil {
+			mErr = multierror.Append(mErr, err)
```

Comment:
`mErr = multierror.Append(mErr, err)`

Reply:
Sorry, can't tell from this comment, what's the suggestion? (Just append without checking err != nil?, or something else?)

Reply:
Think this is addressed now.